### DISCUSSION TITLE
Request Enhancement: -trim

### DISCUSSION THREAD

**Discussion by fmw42** [ImageMagick/ImageMagick#4861] [Development]
It would be nice to be able to limit -trim to some minimum size.  So either an argument to -trim or a -define.

```
-trim <minWidth>x<minHeight>

or

-define trim:minSize=WxH
```

**Comment by urban-warrior** [comment]
The trim geometry is sensitive  to not only the width and height but the offset relative to the image and/or virtual canvas.  It's easy enough to restrict the trim width/height to a minimum bounding box but how is the (x,y) offset attenuated relative to the minimum bounding box for an image with and without a virtual canvas?

**Reply by fmw42** [reply]
How about using it only to limit the trim as centered as possible. That was what I was thinking about when requesting this.

I have never trimmed an image with a virtual canvas. So perhaps ignore it or we can require the virtual canvas to be removed by the user, if they do not want to trim relative to the virtual canvas.

**Comment by urban-warrior** [comment]
We need an example.  Post an image, the trim command-line you use, the results of the trim, a define (-define trim:minSize=WxH), and how the resulting image would change as a result of the define.

**Reply by fmw42** [reply]
Here is an example and some partial code with examples.  The limited trim would need to be gravity relative. So in the following, I give code for center, northwest and southeast. Note that this is just partial code to give an idea of what is being requested.

Create test image (pad rose: image to 256x256).

`convert rose: -background black -gravity center -extent 256x256 rose_256.png`

![rose_256](https://user-images.githubusercontent.com/13458218/157726208-e243d848-0526-4170-a309-32ac6af780a1.png)

Code snippet for testing to limit to 150x150. I just get the trim dimensions and then compute the appropriate crop to limit the trim.

```
minw=150
minh=150
gravity="center"
trimvals=`convert rose_256.png -format "%@" info: | tr "x" "+"`
ww=`echo $trimvals | cut -d+ -f1`
hh=`echo $trimvals | cut -d+ -f2`
xo=`echo $trimvals | cut -d+ -f3`
yo=`echo $trimvals | cut -d+ -f4`
if [ $ww -ge $minw -a $hh -ge $minh ]; then
	convert rose_256.png -crop $trimvals +repage rose_256_trim.png
elif [ $ww -lt $minw -a $hh -lt $minh ]; then
	if [ "$gravity" = "center" ]; then
		diffx=`convert xc: -format "%[fx:round(0.5*($minw-$ww))]" info:`
		diffy=`convert xc: -format "%[fx:round(0.5*($minh-$hh))]" info:`
		xo=$((xo-diffx))
		yo=$((yo-diffy))
		convert rose_256.png -crop ${minw}x${minh}+${xo}+${yo} +repage rose_256_trim_center.png
	elif [ "$gravity" = "northwest" ]; then
		diffx=`convert xc: -format "%[fx:round(($minw-$ww))]" info:`
		diffy=`convert xc: -format "%[fx:round(($minh-$hh))]" info:`
		xo=$((xo-diffx))
		yo=$((yo-diffy))
		convert rose_256.png -crop ${minw}x${minh}+${xo}+${yo} +repage rose_256_trim_nw.png
	elif [ "$gravity" = "southeast" ]; then
		convert rose_256.png -crop ${minw}x${minh}+${xo}+${yo} +repage rose_256_trim_se.png
	fi
fi

```	

Center Result:
![rose_256_trim_center](https://user-images.githubusercontent.com/13458218/157726571-5d479a7b-879e-454c-9cce-8698666a4ff1.png)

Northwest Result:
![rose_256_trim_nw](https://user-images.githubusercontent.com/13458218/157726597-b3d78f66-4481-4264-9a64-d0d2e75998a1.png)

Southeast Result:
![rose_256_trim_se](https://user-images.githubusercontent.com/13458218/157726629-547a00aa-8f2a-4be1-b510-d687b7bb6055.png)

Comments are welcome.  The purpose is not to similar to just trimming and padding back. One may have images that need a fuzzy trim and have some information. Or more likely, it would be needed for some system where the images need to all be of the same size or no smaller than some size.

**Comment by urban-warrior** [comment]
We have the framework for the `minSize` define in place with support for center, northwest, and southeast gravity.  Post the formulations for the remaining gravity types here so we can complete this feature.

**Reply by fmw42** [reply]
Here it is.

[trim.zip](https://github.com/ImageMagick/ImageMagick/files/8358251/trim.zip)


```
minw=150
minh=150
gravity="center"
trimvals=`convert rose_256.png -format "%@" info: | tr "x" "+"`
ww=`echo $trimvals | cut -d+ -f1`
hh=`echo $trimvals | cut -d+ -f2`
xo=`echo $trimvals | cut -d+ -f3`
yo=`echo $trimvals | cut -d+ -f4`
if [ $ww -ge $minw -a $hh -ge $minh ]; then
    convert rose_256.png -crop $trimvals +repage rose_256_trim.png
elif [ $ww -lt $minw -a $hh -lt $minh ]; then
    if [ "$gravity" = "center" ]; then
        diffx=`convert xc: -format "%[fx:round(0.5*($minw-$ww))]" info:`
        diffy=`convert xc: -format "%[fx:round(0.5*($minh-$hh))]" info:`
        xo=$((xo-diffx))
        yo=$((yo-diffy))
        convert rose_256.png -crop ${minw}x${minh}+${xo}+${yo} +repage rose_256_trim_center.png
    elif [ "$gravity" = "northwest" ]; then
        diffx=`convert xc: -format "%[fx:round(($minw-$ww))]" info:`
        diffy=`convert xc: -format "%[fx:round(($minh-$hh))]" info:`
        xo=$((xo-diffx))
        yo=$((yo-diffy))
        convert rose_256.png -crop ${minw}x${minh}+${xo}+${yo} +repage rose_256_trim_nw.png
    elif [ "$gravity" = "north" ]; then
        diffx=`convert xc: -format "%[fx:round(0.5*($minw-$ww))]" info:`
        diffy=`convert xc: -format "%[fx:round(($minh-$hh))]" info:`
        xo=$((xo-diffx))
        yo=$((yo-diffy))
        convert rose_256.png -crop ${minw}x${minh}+${xo}+${yo} +repage rose_256_trim_n.png
    elif [ "$gravity" = "northeast" ]; then
        diffy=`convert xc: -format "%[fx:round(($minh-$hh))]" info:`
        yo=$((yo-diffy))
        convert rose_256.png -crop ${minw}x${minh}+${xo}+${yo} +repage rose_256_trim_ne.png
    elif [ "$gravity" = "east" ]; then
        diffy=`convert xc: -format "%[fx:round(0.5*($minh-$hh))]" info:`
        yo=$((yo-diffy))
        convert rose_256.png -crop ${minw}x${minh}+${xo}+${yo} +repage rose_256_trim_e.png
    elif [ "$gravity" = "southeast" ]; then
        convert rose_256.png -crop ${minw}x${minh}+${xo}+${yo} +repage rose_256_trim_se.png
    elif [ "$gravity" = "south" ]; then
        diffx=`convert xc: -format "%[fx:round(0.5*($minw-$ww))]" info:`
        xo=$((xo-diffx))
        convert rose_256.png -crop ${minw}x${minh}+${xo}+${yo} +repage rose_256_trim_s.png
    elif [ "$gravity" = "southwest" ]; then
        diffx=`convert xc: -format "%[fx:round(($minw-$ww))]" info:`
        xo=$((xo-diffx))
        convert rose_256.png -crop ${minw}x${minh}+${xo}+${yo} +repage rose_256_trim_sw.png
    elif [ "$gravity" = "west" ]; then
        diffx=`convert xc: -format "%[fx:round(($minw-$ww))]" info:`
        diffy=`convert xc: -format "%[fx:round(0.5*($minh-$hh))]" info:`
        xo=$((xo-diffx))
        yo=$((yo-diffy))
        convert rose_256.png -crop ${minw}x${minh}+${xo}+${yo} +repage rose_256_trim_w.png
    fi
fi

```

**Comment by urban-warrior** [comment]
Look for a new beta with a few hours with a patch.  Post here any issue you have with the new define (trim:minSize).

**Reply by fmw42** [reply]
I just downloaded the 7.1.0.30 beta and tried

`magick rose_256.png -define trim:minSize=150x150 -gravity center rose256_c.png`

but the result is still 256x256 and not 150x150, i.e. no change from the input and no trim whatsoever.

**Reply by urban-warrior** [reply]
`-define` and `-gravity` are settings, not image processing operators.  Add `-trim` after `center`.

**Reply by fmw42** [reply]
DOH!  Forgot the main part in my haste

**Reply by fmw42** [reply]
Seems to be working fine, thanks in IM 7.1.0.30 beta

**Reply by LostOnTheLine** [reply]
Is there any way we could get a percentage for the size?
In [another thread](https://github.com/ImageMagick/ImageMagick/discussions/5867) I'm working on a project trying to use this as part of a script. The trim:minSize is great, it gives me what I want so long as it's set to proper resolution, but if set to work with multiple resolutions it'd be nice to have it able to be set with a % of the original size.

My use case, in summary, for more details see the link:

I have many images that have designs such as these that I want to trim borders on

<img src="https://user-images.githubusercontent.com/60684934/206909382-f7e92539-bda0-4147-845d-85f73db4b468.jpg" width="200" /> | <img src="https://user-images.githubusercontent.com/60684934/206909383-2496bf3f-694a-4094-a326-ce5266ee3e1d.jpg" width="200" /> | <img src="https://user-images.githubusercontent.com/60684934/206909384-08969727-bcaa-4aae-be94-17f8ea72919c.jpg" width="200" /> | <img src="https://user-images.githubusercontent.com/60684934/206982758-321161aa-8b01-4c23-a432-62a6b23eb0a1.jpg" width="200" /> | <img src="https://user-images.githubusercontent.com/60684934/207031833-f4515b0a-8d24-4a5d-b5db-40f22e1976d4.jpg" width="200" /> | <img src="https://user-images.githubusercontent.com/60684934/207031838-6c257e45-f25f-4dbb-9cb2-a1dc51ef1c4d.jpg" width="200" />

I want to trim the borders, but leave enough that it still looks alright while keeping the image centered. Ideally I want the trim to be even on all sides as well

With my current working set I need to know the dimensions of the image & an estimated border to set. For the images I'm testing with the original dimensions are `1400x1948` & I found `1200x1700` created a decent border. That's 87% or so so I did a test with 90% & found that setting the geometry as a % did not work, but manually setting it to `1260x1754` gave me a good end result, though unfortunately not centered to the original image, but to the detected borders.




```
magick.exe mogrify -quality 90 -fuzz 1% -gravity Center -define trim:minSize=1260x1754 -trim -filter Lanczos -rotate -90> -strip -resize 800x1100
```

<img src="https://user-images.githubusercontent.com/60684934/207030673-89e1a313-56fb-450a-9167-b2d31c25d9d8.jpg" width="200" /> | <img src="https://user-images.githubusercontent.com/60684934/207030675-76cea3c1-845e-4bb1-b2af-28eb505f8d5b.jpg" width="200" /> | <img src="https://user-images.githubusercontent.com/60684934/207030676-4c70b197-88bd-44da-b7bb-8c74317875fb.jpg" width="200" />


As it doesn't work for my purposes at the moment I'm going to work it into the python script to get the dimension & use them in the call to ImageMagick, but as I feel this is something that was most likely just overlooked I thought I'd report it as a Bug/Request to have it included in the `-define trim:minSize=` argument

**Comment by fmw42** [comment]
I have specified the algorithm such that -gravity identifies the side that is limited. I am wondering if it should be the opposite and identify the side that is trimmed.  Anyone have any strong opinion, otherwise I will leave it as it is.

**Reply by LostOnTheLine** [reply]
I think both have benefits. If defining the amount trimmed it would help with my current project because it would limit the max amount trimmed off of each side essentially making the trimmed image centered to original center instead of to the detected non-border image unless it extended too far out on a given side.

On the contrary the more likely scenario is that someone would want the the resulting image to be centered. Which the current setup favors.

If there is only the ability to have 1 I think the way it currently is is best. But could it be possible to give the ability to choose which to use? That would be ideal In my opinion.
